src/module.c

```diff
@@ -290,7 +290,9 @@ JL_DLLEXPORT jl_binding_t *jl_get_binding_wr(jl_module_t *m JL_PROPAGATES_ROOT,
             if (decode_restriction_kind(pku) != BINDING_KIND_DECLARED) {
                 check_safe_newbinding(m, var);
                 if (!alloc)
-                    jl_errorf("Global %s.%s does not exist and cannot be assigned. Declare it using `global` before attempting assignment.", jl_symbol_name(m->name), jl_symbol_name(var));
+                    jl_errorf("Global %s.%s does not exist and cannot be assigned.\n"
+                              "Note: Julia 1.9 and 1.10 inadvertently omitted this error check (#56933).\n"
+                              "Hint: Declare it using `global` before attempting assignment.", jl_symbol_name(m->name), jl_symbol_name(var));
```

Comment:
```suggestion
                    jl_errorf("Global %s.%s does not exist and cannot be assigned.\n"
                              "Note: Julia 1.9 and 1.10 inadvertently omitted this error check, meaning it was possible to write `{module}.{var} = {value}` without a preceding `:global`.\n"
                              "Hint: Declare it using `global` before attempting assignment:\n"
                              "    ```\n"
                              "    global %s\n"
                              "    %s.%s = ...\n"
                              "    ```",
                              jl_symbol_name(m->name), jl_symbol_name(var),
                              jl_symbol_name(m->name), jl_symbol_name(m->name), jl_symbol_name(var));
```

Reply:
This is just to clarify things, as when I first tried this I thought I could do

```julia
global A.x
```

before realising I needed to evaluate `global` inside `A`.

Reply:
My primary concern with being too explicit here is that there are other paths that reach this error, not just the frontend syntax, so if you reach it through one of those paths, that error will be confusing.

Reply:
I don't think my issue needs to be linked in the error though. Just something to:

1. Help people update their code (the "declare using `global`" should be more precise), and
2. Let them know that this syntax used to work, it wasn't their imagination or mistake

Reply:
`Hint: Declare it using `global %s` inside `%a` before attempting assignment.`

Reply:
> I don't think my issue needs to be linked in the error though.

I think it can be useful. Gives people something to search for and a starting breadcrumb. In the fullness of time, we should have an error database that gives additional information, but that's a big project.

Reply:
Sounds fine to me.

Aside: it would be nice if there was a `Base.declare_global(::Module, sym)` to support this from other modules as otherwise we need to use `Core.eval(::Module, Expr(:global, sym))` which feels a bit dangerous.

Reply:
> which feels a bit dangerous.

Well, it is a dangerous operation, which is one of the big reasons not to do it automatically ;). There may be a future general change to expose all operations currently only available through syntax as intrinsics, but it's not on the immediate roadmap.

Reply:
What makes it dangerous? It will keep working in the future so long as we do `global x`, right? It really is used in a lot of the Python-Julia ecosystem which is the root of my worry.

Reply:
> What makes it dangerous?

In general Julia really doesn't want you to evaluate into modules that are already done. It breaks precompilation, may be slow or cause world age issues. It's allowed for reflection, Revise, etc, but it's not the kind of general purpose dictionary use that python wants to use it for. You're sort of fighting the design here and it's likely to cause additional issues in the future.

Reply:
What module does the REPL evaluate into? Is it `Main`? What are the differences between evaluating into an active module and how the REPL works?

PythonCall.jl is mostly just using this syntax for `Main`. If we could just emulate how the REPL works then maybe we’d be safer long term (?)

Reply:
Thanks. Is there a reason that `Main.x = 1` does not work, even when running within the REPL?

Reply:
Yes, `Main` gets resolved at runtime, while the definitions of globals happens at definition time, when we do not know what the name `Main` refers to. Of course in the REPL, definition time and runtime closely follow each other, but that's not true in general.

Reply:
In particular `f() = (Main.x = 1)` and `f() = (global x = 1)` are not equivalent, because `Main` could be redefined later.

Reply:
Makes sense. I’m just trying to figure out how to use the exact mechanism the REPL uses for this. Because then we can just use that inside PythonCall to communicate new globals.

Reply:
When you write `x = 1`, the REPL does the moral equivalent of `Core.eval(Main, :(x = 1))`. Since there's no Main., the global gets introduced automatically at definition time.

Reply:
Calling `Core.eval(::Module, Expr(:global, sym))` should be perfectly fine here though, as there isn't any opportunity to optimize this code anyways (it is python), and is what `x=1` expands to normally. As Keno mentioned, we should add a builtin too (similar to setglobal, there should be a declareglobal and declareconst functions), so that you don't need the awkward eval form for them, but that is merely a difference in spelling of the operation

Reply:
Got it, thanks for the detailed discussion on this.